src/Google.Datastore.V1Beta3/ArrayValuePartial.cs

```diff
@@ -0,0 +1,82 @@
+        public static implicit operator ArrayValue(DateTimeOffset[] values) => ToArrayValue(values, x => x);
+        public static implicit operator ArrayValue(DateTimeOffset?[] values) => ToArrayValue(values, x => x);
+
+        private static ArrayValue ToArrayValue<T>(T[] values, Func<T, Value> converter) =>
+            values == null ? null : new ArrayValue { Values = { values.Select(x => converter(x) ?? Value.ForNull()) } };
+
+        public static explicit operator string[](ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (string) x);
+        public static explicit operator long[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (long)x);
+        public static explicit operator long?[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (long?)x);
+        public static explicit operator Key[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (Key)x);
+        public static explicit operator double[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (double)x);
+        public static explicit operator double?[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (double?)x);
+        public static explicit operator bool[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (bool)x);
+        public static explicit operator bool?[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (bool?)x);
+        public static explicit operator ArrayValue[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (ArrayValue)x);
+        public static explicit operator ByteString[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (ByteString)x);
+        public static explicit operator byte[][] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (byte[])x);
+        public static explicit operator Entity[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (Entity)x);
+        public static explicit operator LatLng[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (LatLng)x);
+        public static explicit operator Timestamp[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (Timestamp)x);
+        public static explicit operator DateTime[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (DateTime)x);
+        public static explicit operator DateTime?[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (DateTime?)x);
+        public static explicit operator DateTimeOffset[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (DateTimeOffset)x);
+        public static explicit operator DateTimeOffset?[] (ArrayValue arrayValue) => FromArrayValue(arrayValue, x => (DateTimeOffset?)x);
+
+        private static T[] FromArrayValue<T>(ArrayValue arrayValue, Func<Value, T> converter) =>
+            arrayValue == null ? null : arrayValue.Values.Select(x => converter(x)).ToArray();
+
+        public IEnumerator<Value> GetEnumerator() => Values.Select(x => x.OrNull()).GetEnumerator();
```

Comment:
Why not just return the enumerator for the values directly? Is the thought here that while iterating over an ArrayValue, checking for null would be cleaner than `value.ValueTypeCase != Value.ValueTypeOneofCase.NullValue`? It feels a little odd to me that iterating over an ArrayValue behaves differently than iterating over it Values collection, but not enough to forgo cleaner null checks.

If this is the motivation, perhaps a cleaner way of checking for null, such as an IsNull property on Value, would be something to consider.

By the way, going off on a tangent here, but it looks like Value.NullValue always returns the same value: if the value is a null, it returns `Google.Protobuf.WellKnownTypes.NullValue.NullValue`, which is the 0 value, or it returns 0. Seems like a property which is only useful to set. I assume there's not much that can be done here since it's auto-generated, right?